### PR TITLE
sql: SHOW TABLES marks a dropped table with a " (dropped)" suffix

### LINKED ISSUE #9318: sql: DROP does not properly mark tables as deleted
- What did you do?

dropped a table.

- What did you expect to see?

root@178.62.29.223:26257> show tables;
```
+-------+
| Table |
+-------+
| logs  |
+-------+
(1 row)
root@178.62.29.223:26257> drop table logs;
DROPPED
root@178.62.29.223:26257> show tables;
+-------+
| Table |
+-------+
+-------+
(0 rows)
```

- What did you see instead?

```
root@178.62.29.223:26257> show tables;
+-------+
| Table |
+-------+
| logs  |
+-------+
(1 row)
root@178.62.29.223:26257> drop table logs;
pq: table "logs" has been deleted
root@178.62.29.223:26257> show tables;
+-------+
| Table |
+-------+
| logs  |
+-------+
(1 row)
```

about the node:
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] file created at: 2016/09/13 13:50:53
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] running on machine: cdb1
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] binary: CockroachDB beta-20160908 (linux amd64, built 2016/09/08 15:33:58, go1.7)
I160913 13:50:53.910615 1 util/log/clog.go:1019  [config] arguments: [cockroach start --host=178.62.29.223 --http-addr=127.0.0.1 --ca-cert=ca.cert --cert=node1.cert --key=node1.key]
I160913 13:50:53.910615 1 util/log/clog.go:1019  line format: [IWEF]yymmdd hh:mm:ss.uuuuuu goid file:line msg
I160913 13:50:53.910607 1 cli/start.go:342  CockroachDB beta-20160908 (linux amd64, built 2016/09/08 15:33:58, go1.7)

### PR DESCRIPTION
**Pull request by vivekmenezes**
fixes #9318

Added another commit that renames table deleted to table dropped for #9493

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/10063)

### CODE REVIEW

**Comment by dt** [comment]
hm, how do we feel about `show tables` and the introspection meta tables (`pg_*` and `information_schema`) being different?

**Comment by dt** [comment]
I wonder if we should just change desc.Name in-place when we flip the Deleted bit, so all the paths read the same thing

**Comment by vivekmenezes** [comment]
I'm not familiar with what you're talking about. What's broken here?

**Comment by dt** [comment]
in addition to `SHOW TABLES`, we support reflecting on tables using the metatables in `information_schema` or `pg_catalog`. Having `show tables` show one name while they show something else is potentially confusing (plus there's a goal to unify their implementation, potentially replacing the `SHOW` commands with queries into the meta tables). While this isn't broken yet, it just introduces inconsistency when it's only implemented here.

**Comment by knz** [comment]
I'd recommend linking this with https://github.com/cockroachdb/cockroach/issues/9966.

File: pkg/sql/lease_test.go [hunk 1]
```diff
@@ -454,6 +455,23 @@ CREATE TABLE test.t(a INT PRIMARY KEY);
 	if !testutils.IsError(err, "table is being deleted") {
 		t.Fatalf("got a different error than expected: %v", err)
 	}
+
+	// Check that SHOW TABLES marks a deleted table with the " (deleted)"
+	// suffix.
+	rows, err := t.db.Query(`SHOW TABLES FROM test`)
```
Comment (andreimatei) [inline pkg/sql/lease_test.go]: defer `rows.Close()`

File: pkg/sql/lease_test.go [hunk 2]
```diff
@@ -401,7 +401,8 @@ func TestLeaseManagerPublishVersionChanged(testingT *testing.T) {
 	t.expectLeases(descID, "/3/1")
 }
 
-// Test that we fail to lease a table that was marked for deletion.
+// Test that we fail to lease a table that was marked for deletion. This also
+// tests that SHOW TABLES marks a deleted table with the " (deleted)" suffix.
```
Comment (andreimatei) [inline pkg/sql/lease_test.go]: I think it'd be more natural for this to be a test on `SHOW TABLES`

**Comment by andreimatei** [comment]
LGTM

this looks to me like a quick fix for something that has come up many times. Whether or not the table should appear in information schema is a valid question; I'd say leave it alone there, since we probably shouldn't be massaging names like we're doing in the `SHOW` output.
Once we back the implementation of `SHOW TABLES` with queries in info schema, we can still continue doing this massaging, I think that's fair.

File: pkg/sql/lease_test.go (hunk 2 above)
Comment (vivekmenezes) [inline pkg/sql/lease_test.go]: Done. Moved to show_test.go

File: pkg/sql/lease_test.go (hunk 1 above)
Comment (vivekmenezes) [inline pkg/sql/lease_test.go]: Done

**Comment by vivekmenezes** [comment]
Any thoughts on this PR?

**Comment by andreimatei** [comment]
Still LGTM as far as I'm concerned.
Was there a particular point to r3? If not, there is a cost to churn...

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_test.go, line 502 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063#-KV6TmSp3O7n29j-oRwp:-KV6TmSp3O7n29j-oRwq:bvwjbjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ba8b97362d0c3373b558540b962a2edaaef0a5a/pkg/sql/drop_test.go#L502)):*
> ```Go
> 		t.Fatal("table invisible through SHOW TABLES")
> 	}
> 	var val []byte
> ```

can't this be a string?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063#-:-KV6TsO2-yLz4dq0HgE2:bkhzbxj)*

**Comment by vivekmenezes** [comment]
The dropped versus deleted incompatibility was driving me crazy!

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/drop_test.go, line 502 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063#-KV6TmSp3O7n29j-oRwp:-KV6YE4BBLUy5XkgSh8H:b-hxpjjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ba8b97362d0c3373b558540b962a2edaaef0a5a/pkg/sql/drop_test.go#L502)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can't this be a string?
</details>
Yes it can. I was under the impression that a string was immutable: https://golang.org/ref/spec#String_types

I'm puzzled by what they mean by immutable

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063#-:-KV6YvEfB3Bs3lZiIHZF:b-rhk176)*

**Comment by andreimatei** [comment]
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_test.go, line 502 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063#-KV6TmSp3O7n29j-oRwp:-KV6eqVOA2VajimyWYQH:bvowprv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8ba8b97362d0c3373b558540b962a2edaaef0a5a/pkg/sql/drop_test.go#L502)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> Yes it can. I was under the impression that a string was immutable: https://golang.org/ref/spec#String_types
> 
> I'm puzzled by what they mean by immutable
</details>
the mean that you can't change the contents of a string instance. Here you're passing a pointer to a string to `Scan`, and the instance being pointed to will be overwritten by a completely new instance.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063)*

**Comment by tamird** [comment]
Reviewed 9 of 9 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/drop_test.go, line 470 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063#-KV6ynqkcrpDDXB4gSIv:-KV6ynqkcrpDDXB4gSIw:b-evte55) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb636d5227e0ed397893158b9ab5ec1b45757d4/pkg/sql/drop_test.go#L470)):*
> ```Go
> 	params.Knobs = base.TestingKnobs{
> 		SQLSchemaChanger: &sql.SchemaChangerTestingKnobs{
> 			SyncFilter: func(tscc sql.TestingSchemaChangerCollection) {
> ```

FYI: this can be written as `SyncFilter: (*sql.TestingSchemaChangerCollection).ClearSchemaChangers`. Just sharing.

---

*[pkg/sql/show.go, line 668 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063#-KV6nYahD5j3Sibnr6T6:-KV6nYahD5j3Sibnr6T7:b-d41gx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/cbb636d5227e0ed397893158b9ab5ec1b45757d4/pkg/sql/show.go#L668)):*
> ```Go
> 				tableName := name.Table()
> 				// Check to see if the table has been dropped.
> 				if _, err := p.mustGetTableOrViewDesc(&name); err != nil {
> ```

NYC but a function called `mustGet...` returns an error? What has the world come to?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10063)*

**Pull request merged**